### PR TITLE
Ton of refactors + 3D Manipulation

### PR DESCRIPTION
**Pull request by praveenVnktsh**
Uatu created: 
- Manages all perception tracker tasks. 
- Uatu contains methods for tracking objects in the scene during runtime, generating volumetric representations of the scene, obtaining point clouds and reconstructing objects in the scene.

Uatu is currently a python package within manipulation since it is not used by any other methods outside manip. It can be pushed to a separate ros package in the event of need by other packages.

3D Manipulation
- Has methods for planning for base and manipulator tasks.

Updates to old code - 
- Object detection is now publishing detections based on input images only. Detections are rotated to match the horizontal input images. This makes everything easier to handle downstream

- Implicitly tracks all objects in the scene, and maintains a Bag of Words representation for LLM inferencing
- Maintains an explicit tracker for objects of interest with ability to perform 3D reconstruction of a specific object from multiple views, and multiple base poses of the robot. Can be directly used for high quality grasp generation using graspnet + 3D manipulation
- New instance disambiguation technique for object detections using distances between distributions. 
- Contains methods for 3D volumetric reconstruction, object tracking and scene parsing
- Adds extensive infrastructure to quickly add new kinds of planners on 3D clouds . 
- Implements naive planner, a_star based planning, and reachability planning for base

### CODE REVIEW

File: src/common/alfred_core/scripts/alfred_driver.py [hunk 1]
```diff
@@ -49,6 +50,18 @@ def __init__(self):
...
+            'intrinsics' : None,
+        }
+        
+        
+        # camera services
+        
+        
+        
+        
+
```
Comment (atharva-18) [inline src/common/alfred_core/scripts/alfred_driver.py]:
```suggestion
        # camera services
```

File: src/manipulation/scripts/debug/collect_data.py [hunk 2]
```diff
@@ -0,0 +1,267 @@
...
+        rospy.sleep(2)
+        nRotates = 0
+        for i, angle in enumerate(np.arange(self.vs_range[0][0], self.vs_range[0][1], self.vs_range[1])):
+            move_to_pose(self.trajectoryClient, {
+                'head_pan;to' : angle,
+            })
+            rospy.sleep(1.5) # wait for the head to move
+            cv2.imwrite(f'/home/hello-robot/alfred-autonomy/src/manipulation/scripts/debug/images/rgb/{str(i).zfill(6)}.png', self.color_image)
+            cv2.imwrite(f'/home/hello-robot/alfred-autonomy/src/manipulation/scripts/debug/images/depth/{str(i).zfill(6)}.tif', self.depth_image)
+            with open(f'/home/hello-robot/alfred-autonomy/src/manipulation/scripts/debug/images/pose/{str(i).zfill(6)}.json', 'w') as f:
```
Comment (atharva-18) [inline src/manipulation/scripts/debug/collect_data.py:146]: Can we replace these with `os.path.join` and `os.expanduser`

File: src/manipulation/scripts/new/helpers.py [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+        joint_efforts = [pose[key][1] for key in joint_names]
+        point.positions = joint_positions
+        point.effort = joint_efforts
+        trajectory_goal.trajectory.points = [point]
+    trajectory_goal.trajectory.header.stamp = rospy.Time.now()
+    trajectoryClient.send_goal(trajectory_goal)
+    if not asynchronous: 
+        trajectoryClient.wait_for_result()
+        #print('Received the following result:')
+        #print(self.trajectory_client.get_result())
```
Comment (atharva-18) [inline src/manipulation/scripts/new/helpers.py]:
```suggestion
```

File: src/manipulation/scripts/new/helpers.py [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+
+
+
+if __name__ == "__main__":
+    # unit tests
+    img = np.random.rand(500, 500)
+    vals= convertPointToDepth(img, np.array([[1,2],[3,4]]), {'fx': 1, 'fy': 1, 'ppx': 1, 'ppy': 1})
+    # point[2] = depth
```
Comment (atharva-18) [inline src/manipulation/scripts/new/helpers.py:77]:
```suggestion
```

File: src/manipulation/scripts/new/manipulation_new.py [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+        self.node_name = "manipulation"
+        rospy.init_node(self.node_name, anonymous=True)
+        
+        
+        rospy.loginfo(f"[{self.node_name}]: Waiting for services...")
+        # self.stow_robot_service = rospy.ServiceProxy('/stow_robot', Trigger)
+        # self.stow_robot_service.wait_for_service()
+        
+        # self.navigate_to_location_service = rospy.ServiceProxy('/navigate_to_location', Trigger)
+        # self.navigate_to_location_service.wait_for_service()
```
Comment (atharva-18) [inline src/manipulation/scripts/new/manipulation_new.py:23]:
```suggestion
```

File: src/manipulation/scripts/new/manipulation_new.py [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+        #     curvel = vx
+        #     prevsettime = time.time()
+            
+        #     rospy.sleep(0.1)
+        # move_to_pose(self.trajectoryClient, {
+        #     'base_translate;vel' : 0.0,
+        # }) 
+        # risk
+        # if self.isDetected is False:
+        #     return False
```
Comment (atharva-18) [inline src/manipulation/scripts/new/manipulation_new.py:122]: TODO

File: src/manipulation/scripts/new/planner/planner.py [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+
+        img = cv2.circle(img, (base_loc[0], base_loc[1])[::-1], 1, (0, 255, 255), -1)
+
+
+        img = cv2.circle(img, convert_point_to_img(source[0], source[1])[::-1], 2, (0, 255, 0), -1)
+        img = cv2.circle(img, convert_point_to_img(target[0], target[1])[::-1], self.radius_reachable, (255, 0, 0), 1)
+        img = cv2.circle(img, convert_point_to_img(target[0], target[1])[::-1], 2, (255, 0, 0), -1)
+        img = cv2.circle(img, convert_point_to_img(source[0], source[1])[::-1], self.max_dist_traverse, (0, 255, 0), 1)
+
+        cv2.imwrite("/home/praveenvnktsh/alfred-autonomy/src/manipulation/scripts/new/reachability_map.png", img)
```
Comment (atharva-18) [inline src/manipulation/scripts/new/planner/planner.py:102]: Hardcoded path

File: src/manipulation/scripts/new/uatu/parse_scene.py [hunk 8]
```diff
@@ -0,0 +1,245 @@
...
+from yolo.msg import Detections
+from uatu.graph import Graph, Node
+import message_filters
+import open3d as o3d
+
+
+class SceneParser:
+    def __init__(self,):
+        self.node_name = 'uatu'
+        # rospy.init_node(self.node_name, anonymous=True)
```
Comment (atharva-18) [inline src/manipulation/scripts/new/uatu/parse_scene.py:25]:
```suggestion
```

File: src/manipulation/scripts/new/uatu/parse_scene.py [hunk 9]
```diff
@@ -0,0 +1,245 @@
...
+
+        self.bridge = CvBridge()
+        
+
+        self.objectLocArr = []
+
+        self.depth_image = None
+        self.color_image = None
+
+        # depth_topic = '/camera/aligned_depth_to_color/image_raw'
```
Comment (atharva-18) [inline src/manipulation/scripts/new/uatu/parse_scene.py:50]:
```suggestion
```

File: src/manipulation/scripts/new/uatu/parse_scene.py [hunk 10]
```diff
@@ -0,0 +1,245 @@
...
+
+        self.depth_image_subscriber = message_filters.Subscriber(depth_topic, Image)
+        self.rgb_image_subscriber = message_filters.Subscriber(color_topic, Image)
+        synced = message_filters.ApproximateTimeSynchronizer([self.depth_image_subscriber, self.rgb_image_subscriber], 10, 0.1, allow_headerless=True)
+        synced.registerCallback(self.image_callback)
+        
+        self.boundingBoxSub = rospy.Subscriber("/object_bounding_boxes", Detections, self.parse_detections)
+        
+        # self.class_list = rospy.get_param('class_list')
+        # self.graph : Graph = Graph(len(self.class_list))
```
Comment (atharva-18) [inline src/manipulation/scripts/new/uatu/parse_scene.py:62]:
```suggestion
```

File: src/manipulation/scripts/new/uatu/parse_scene.py [hunk 11]
```diff
@@ -0,0 +1,245 @@
...
+            sdf_trunc=0.04,
+            color_type=o3d.pipelines.integration.TSDFVolumeColorType.RGB8
+        )
+        return True, "Accumulated TSDF"
+    
+    def stop_accumulate_tsdf(self, req = None):
+        self.accumulating_tsdf = False
+        pcd = self.tsdf_volume.extract_point_cloud()
+        self.point_cloud = pcd
+        # path_to_tsdf = "/home/hello-robot/alfred-autonomy/src/perception/uatu/outputs/tsdf.ply"
```
Comment (atharva-18) [inline src/manipulation/scripts/new/uatu/parse_scene.py:105]:
```suggestion
```

File: src/manipulation/scripts/new/uatu/parse_scene.py [hunk 12]
```diff
@@ -0,0 +1,245 @@
...
+        path_to_tsdf = "/home/praveenvnktsh/alfred-autonomy/src/perception/uatu/outputs/tsdf.ply"
```
Comment (atharva-18) [inline src/manipulation/scripts/new/uatu/parse_scene.py:106]: Hardcoded path

File: src/manipulation/scripts/new/uatu/parse_scene.py [hunk 13]
```diff
@@ -0,0 +1,245 @@
...
+        return True
+
+    def image_snapshot(self):
+        camera_frame_name = '/camera_color_optical_frame'
+        
+        extrinsics_to_base = self.get_transform(camera_frame_name, '/base_link')
+        extrinsics_to_base = np.concatenate((extrinsics_to_base, np.array([[0, 0, 0, 1]])), axis=0)
+        extrinsics_to_map = extrinsics_to_base
+        # extrinsics_to_map = self.get_transform(camera_frame_name, '/map')
+        # extrinsics_to_map = np.concatenate((extrinsics_to_map, np.array([[0, 0, 0, 1]])), axis=0)
```
Comment (atharva-18) [inline src/manipulation/scripts/new/uatu/parse_scene.py:152]:
```suggestion
```

File: src/manipulation/scripts/new/uatu/parse_scene.py [hunk 14]
```diff
@@ -0,0 +1,245 @@
...
+        self.depth_image = np.array(depth_img, dtype=np.uint16)
+        self.color_image = self.bridge.imgmsg_to_cv2(color_img, desired_encoding="passthrough")
+        self.is_receiving_images = True
+            
+    def get_three_d_bbox_from_tracker(self):
+        if self.trackedNode == None or self.point_cloud == None:
+            return False
+        
+        cloud = self.trackedNode.reconstruct_object()
+        # o3d.visualization.draw_geometries([cloud])
```
Comment (atharva-18) [inline src/manipulation/scripts/new/uatu/parse_scene.py]:
```suggestion
```

File: src/manipulation/scripts/new/uatu/parse_scene.py [hunk 15]
```diff
@@ -0,0 +1,245 @@
...
+        #     x1, y1, x2, y2 =  box
+        #     mask = np.zeros_like(self.depth_image)
+        #     mask[int(y1):int(y2), int(x1):int(x2)] = 1
+        #     depth = self.depth_image * mask
+        #     color = self.color_image * mask
+            
+        #     self.graph.add_observation(
+        #         color, depth,
+        #         class_type, confidence, 
+        #         msg.header.stamp)
```
Comment (atharva-18) [inline src/manipulation/scripts/new/uatu/parse_scene.py:247]:
```suggestion
```

File: src/manipulation/scripts/new/view_mesh.py [hunk 16]
```diff
@@ -0,0 +1,5 @@
+import open3d as o3d
+
+
+pcd = o3d.io.read_point_cloud("/home/praveenvnktsh/alfred-autonomy/src/perception/uatu/outputs/tsdf.ply")
```
Comment (atharva-18) [inline src/manipulation/scripts/new/view_mesh.py]: Hardcoded path

File: src/manipulation/scripts/sim/fusion.py [hunk 17]
```diff
@@ -0,0 +1,431 @@
...
+  def vox2world(vol_origin, vox_coords, vox_size):
+    """Convert voxel grid coordinates to world coordinates.
+    """
+    vol_origin = vol_origin.astype(np.float32)
+    vox_coords = vox_coords.astype(np.float32)
+    cam_pts = np.empty_like(vox_coords, dtype=np.float32)
+    cam_pts = vol_origin + (vox_size * vox_coords)
+    # for i in prange(vox_coords.shape[0]):
+    #   for j in range(3):
+    #     cam_pts[i, j] = vol_origin[j] + (vox_size * vox_coords[i, j])
```
Comment (atharva-18) [inline src/manipulation/scripts/sim/fusion.py:179]:
```suggestion
```

File: src/manipulation/scripts/sim/fusion.py [hunk 18]
```diff
@@ -0,0 +1,431 @@
...
+    """
+    intr = intr.astype(np.float32)
+    fx, fy = intr[0, 0], intr[1, 1]
+    cx, cy = intr[0, 2], intr[1, 2]
+    pix = np.empty((cam_pts.shape[0], 2), dtype=np.int64)
+    pix[:, 0] = np.round((cam_pts[:, 0] * fx / cam_pts[:, 2]) + cx)
+    pix[:, 1] = np.round((cam_pts[:, 1] * fy / cam_pts[:, 2]) + cy)
+    # for i in prange(cam_pts.shape[0]):
+    #   pix[i, 0] = int(np.round((cam_pts[i, 0] * fx / cam_pts[i, 2]) + cx))
+    #   pix[i, 1] = int(np.round((cam_pts[i, 1] * fy / cam_pts[i, 2]) + cy))
```
Comment (atharva-18) [inline src/manipulation/scripts/sim/fusion.py:195]:
```suggestion
```

File: src/manipulation/scripts/sim/fusion.py [hunk 19]
```diff
@@ -0,0 +1,431 @@
...
+  def integrate_tsdf(tsdf_vol, dist, w_old, obs_weight):
+    """Integrate the TSDF volume.
+    """
+    tsdf_vol_int = np.empty_like(tsdf_vol, dtype=np.float32)
+    w_new = np.empty_like(w_old, dtype=np.float32)
+    w_new = w_old + obs_weight
+    tsdf_vol_int = (w_old * tsdf_vol + obs_weight * dist) / w_new
+    # for i in prange(len(tsdf_vol)):
+    #   w_new[i] = w_old[i] + obs_weight
+    #   tsdf_vol_int[i] = (w_old[i] * tsdf_vol[i] + obs_weight * dist[i]) / w_new[i]
```
Comment (atharva-18) [inline src/manipulation/scripts/sim/fusion.py:209]:
```suggestion
```

File: src/manipulation/scripts/sim/generate_height_plan.py [hunk 20]
```diff
@@ -0,0 +1,123 @@
+import numpy as np
+import open3d as o3d 
+import cv2
+pcd = o3d.io.read_point_cloud("/home/praveenvnktsh/alfred-autonomy/src/manipulation/scripts/sim/pcd.ply")
```
Comment (atharva-18) [inline src/manipulation/scripts/sim/generate_height_plan.py:4]: Hardcoded path

File: src/manipulation/scripts/sim/generate_height_plan.py [hunk 21]
```diff
@@ -0,0 +1,123 @@
...
+y_min = np.min(y_proj)
+y_max = np.max(y_proj)
+
+print(np.mean(x_proj), np.mean(y_proj))
+
+source = np.array([0.0, 0.0])
+target = np.array([0.8, -0.1])
+theta = 0
+theta = theta * np.pi/180
+# target = source + 1.5 * np.array([np.cos(theta), np.sin(theta)])
```
Comment (atharva-18) [inline src/manipulation/scripts/sim/generate_height_plan.py:20]:
```suggestion
```

File: src/manipulation/scripts/sim/generate_height_plan.py [hunk 22]
```diff
@@ -0,0 +1,123 @@
...
+def convert_point(x, y):
+    x_ind = int((x - x_min)/resolution)
+    y_ind = int((y - y_min)/resolution)
+    return x_ind, y_ind
+    
+
+for point in sorted_points:
+    x, y, z = point
+    # x_ind = int((x - x_min)/resolution)
+    # y_ind = int((y - y_min)/resolution)
```
Comment (atharva-18) [inline src/manipulation/scripts/sim/generate_height_plan.py:40]:
```suggestion
```

File: src/manipulation/scripts/sim/generate_height_plan.py [hunk 23]
```diff
@@ -0,0 +1,123 @@
...
+
+indices = (safety_dist > 0.5) & (safety_dist < 0.6)
+
+reachability_map = np.zeros_like(img[:, :, 0], dtype = np.uint8)
+source_reachability_map = np.zeros_like(reachability_map)
+
+cv2.circle(reachability_map, convert_point(target[0], target[1])[::-1], radius_reachable, 255, -1)
+cv2.circle(source_reachability_map, convert_point(source[0], source[1])[::-1], max_dist_traverse, 255, -1)
+
+# reachability_map = ((reachability_map & binary) * 255).astype(np.uint8)
```
Comment (atharva-18) [inline src/manipulation/scripts/sim/generate_height_plan.py:81]:
```suggestion
```

File: src/manipulation/scripts/sim/generate_height_plan.py [hunk 24]
```diff
@@ -0,0 +1,123 @@
...
+
+characteristic_dist = 0.15 / resolution
+target_reachable_dist = cv2.distanceTransform(reachability_map, cv2.DIST_L2, 5)
+source_reachabile_dist = cv2.distanceTransform(source_reachability_map, cv2.DIST_L2, 5)
+print(np.max(target_reachable_dist), np.max(source_reachabile_dist), characteristic_dist)
+# target_reachable_dist[target_reachable_dist < characteristic_dist] = 0
+# source_reachabile_dist[source_reachabile_dist < characteristic_dist] = 0
```
Comment (atharva-18) [inline src/manipulation/scripts/sim/generate_height_plan.py:88]:
```suggestion
```

File: src/manipulation/scripts/sim/generate_tsdf_from_images.py [hunk 25]
```diff
@@ -0,0 +1,59 @@
...
+    #     classes = np.array(detections['classes'])
+    #     bottles = classes == 39
+    #     boxes = boxes[bottles]
+    #     if len(boxes) == 0:
+    #         depthimg = np.zeros_like(depthimg)
+    #     for box in boxes:
+    #         depthmask = np.zeros_like(depthimg)
+    #         depthmask[int(box[1]):int(box[3]), int(box[0]):int(box[2])] = 1
+    #         depthimg = depthimg * depthmask
+    #         print(depthimg.shape)
```
Comment (atharva-18) [inline src/manipulation/scripts/sim/generate_tsdf_from_images.py:28]:
```suggestion
```

File: src/manipulation/scripts/sim/generate_tsdf_from_images.py [hunk 26]
```diff
@@ -0,0 +1,59 @@
...
+    cam.intrinsic_matrix = intrinsics
+    cam.width = colorimg.shape[1]
+    cam.height = colorimg.shape[0]
+    volume.integrate(
+        rgbd,
+        cam,
+        extrinsics
+    )
+mesh = volume.extract_point_cloud()
+# mesh = volume.extract_triangle_mesh()
```
Comment (atharva-18) [inline src/manipulation/scripts/sim/generate_tsdf_from_images.py:57]:
```suggestion
```

File: src/manipulation/scripts/sim/generate_tsdf_from_images.py [hunk 27]
```diff
@@ -0,0 +1,59 @@
...
+o3d.io.write_point_cloud("/home/praveenvnktsh/alfred-autonomy/src/manipulation/scripts/sim/pcd.ply", mesh)
```
Comment (atharva-18) [inline src/manipulation/scripts/sim/generate_tsdf_from_images.py:58]: Hardcoded path

File: src/manipulation/scripts/sim/generate_tsdf_from_images.py [hunk 28]
```diff
@@ -0,0 +1,59 @@
...
+# o3d.visualization.draw_geometries([mesh])
```
Comment (atharva-18) [inline src/manipulation/scripts/sim/generate_tsdf_from_images.py:59]:
```suggestion
```

File: src/manipulation/scripts/sim/perception_new.py [hunk 29]
```diff
@@ -145,33 +145,67 @@ def aggregateCloudIntoOccupancy(self):
...
+            # fuser.integrate(color, depth, intrinsics, extrinsics, weight)
+            # self.getPointcloud()
+            # self.aggregateCloudIntoOccupancy(
+        # tsdf, col = fuser.get_volume()
+        # np.save("tsdf.npy", tsdf)
+        # np.save("col.npy", col)
+        # point_cloud = fuser.get_point_cloud()
+        # # fusion.meshwrite("mesh.ply", verts, faces, norms, colors)
+        # fusion.pcwrite("pc.ply", point_cloud)
+        # plot_plane(self.occupancyGrid, self.voxelResolution)
```
Comment (atharva-18) [inline src/manipulation/scripts/sim/perception_new.py:202]:
```suggestion
```

File: src/manipulation/scripts/sim/perception_new.py [hunk 30]
```diff
@@ -145,33 +145,67 @@ def aggregateCloudIntoOccupancy(self):
...
+        mesh = volume.extract_triangle_mesh()
+        mesh.compute_vertex_normals()
+        o3d.visualization.draw_geometries([mesh],
+                                      )
+        o3d.io.write_triangle_mesh("mesh.ply", mesh)
+        # np.save("occupancyGrid_005.npy", self.occupancyGrid)  # 0.1m resolution and 0.05
```
Comment (atharva-18) [inline src/manipulation/scripts/sim/perception_new.py:208]:
```suggestion
```

File: src/manipulation/scripts/sim/perception_new.py [hunk 31]
```diff
@@ -256,8 +290,16 @@ def plot_plane(occupancyGrid, resolution):
...
-    objectLocalizer.collectData()
+    print("Starting to collect data")
+    from fusion import TSDFVolume
+    vol_bnds = np.zeros((3,2))
+    vol_bnds[:, 0] = np.array([-2, -2, 0])
+    vol_bnds[:, 1] = np.array([2, 2, 3])
+    fuser = TSDFVolume(vol_bnds, voxel_size=0.05)
+    
+    objectLocalizer.collectData(fuser)
+    # fuser.integrate(objectLocalizer.points, np.eye(4))
```
Comment (atharva-18) [inline src/manipulation/scripts/sim/perception_new.py:301]:
```suggestion
```

File: src/manipulation/scripts/sim/planner_with_omp.py [hunk 32]
```diff
@@ -0,0 +1,50 @@
...
+from ompl import geometric as og
+ 
+def isStateValid(state):
+    return state.getX() < .6
+ 
+def plan():
+    # create an SE2 state space
+    space = ob.SE2StateSpace()
+    # newspace = ob.CompoundStateSpace()
+    # newspace.addSubspace(space, 1.0)
```
Comment (atharva-18) [inline src/manipulation/scripts/sim/planner_with_omp.py:12]:
```suggestion
```

File: src/manipulation/scripts/sim/predict_yolo.py [hunk 33]
```diff
@@ -0,0 +1,57 @@
+from ultralytics import YOLO
+import glob
+import os
+import cv2
+import supervision as sv
+import json
+import numpy as np
+
+os.makedirs('/home/praveenvnktsh/alfred-autonomy/src/manipulation/scripts/sim/images/detections', exist_ok=True)
+os.makedirs('/home/praveenvnktsh/alfred-autonomy/src/manipulation/scripts/sim/images/detection_img', exist_ok=True)
```
Comment (atharva-18) [inline src/manipulation/scripts/sim/predict_yolo.py:10]: Hardcoded path

File: src/manipulation/scripts/sim/predict_yolo.py [hunk 34]
```diff
@@ -0,0 +1,57 @@
...
+# Load a model
+model = YOLO("yolov8n.pt")  # load a pretrained model (recommended for training)
+
+
+for i, path in enumerate(sorted(glob.glob('/home/praveenvnktsh/alfred-autonomy/src/manipulation/scripts/sim/images/rgb/*.png'))):
```
Comment (atharva-18) [inline src/manipulation/scripts/sim/predict_yolo.py:15]: Hardcoded path

File: src/manipulation/scripts/sim/predict_yolo.py [hunk 35]
```diff
@@ -0,0 +1,57 @@
...
+        final_results[0].append(b)
+        final_results[1].append(cls) #remapped class index
+        final_results[2].append(conf)
+        img = cv2.rectangle(img, (int(b[0]),int(b[1])),(int(b[2]),int(b[3])) , (255,0,0), 2)
+        
+        img = cv2.putText(img, str(cls) ,(int(b[0]),int(b[1])),  cv2.FONT_HERSHEY_SIMPLEX, 
+                    0.5, (0,0,255), 1, cv2.LINE_AA)
+        
+    cv2.imwrite('/home/praveenvnktsh/alfred-autonomy/src/manipulation/scripts/sim/images/detection_img/' + str(i).zfill(6) + '.png' , img)
+    with open('/home/praveenvnktsh/alfred-autonomy/src/manipulation/scripts/sim/images/detections/' + str(i).zfill(6)  + '.json', 'w') as f:
```
Comment (atharva-18) [inline src/manipulation/scripts/sim/predict_yolo.py:51]: Hardcoded paths

File: src/manipulation/scripts/sim/viztsdf.py [hunk 36]
```diff
@@ -0,0 +1,12 @@
+# #!/usr/bin/env python3
+# import open3d as o3d
+# import numpy as np
+# import matplotlib.pyplot as plt
+# tsdf = np.load( "/home/praveenvnktsh/alfred-autonomy/src/manipulation/scripts/sim/tsdf.npy" )
+# print(tsdf.shape)
+# o3d.integration.TSDFVolume()
+# # visualize tsdf
+# plt.imshow(tsdf[:,:,0])
+# plt.show()
```
Comment (atharva-18) [inline src/manipulation/scripts/sim/viztsdf.py:10]: Do we need this?

File: src/manipulation/scripts/threed_planning/inverse_kinematics.py [hunk 37]
```diff
@@ -0,0 +1,49 @@
+import numpy as np
+import ikpy.urdf.utils
+import matplotlib.pyplot as plt
+import networkx as nx
+iktuturdf_path = '/home/praveenvnktsh/alfred-autonomy/src/common/alfred_core/config/stretch_ik.urdf'
+tree = ikpy.urdf.utils.get_urdf_tree(iktuturdf_path, "base_link")[0]
+# display.display_png(tree)
+# nx.draw_networkx(tree)
+# plt.show()
+# print(tree)
```
Comment (atharva-18) [inline src/manipulation/scripts/threed_planning/inverse_kinematics.py:10]:
```suggestion
```

File: src/manipulation/scripts/threed_planning/inverse_kinematics.py [hunk 38]
```diff
@@ -0,0 +1,49 @@
+import numpy as np
+import ikpy.urdf.utils
+import matplotlib.pyplot as plt
+import networkx as nx
+iktuturdf_path = '/home/praveenvnktsh/alfred-autonomy/src/common/alfred_core/config/stretch_ik.urdf'
```
Comment (atharva-18) [inline src/manipulation/scripts/threed_planning/inverse_kinematics.py:5]: Hardcoded path

**Comment by praveenVnktsh** [comment]
Accidentally made some changes in alfred_driver to this branch. keeping it here for now :(

Updates  https://www.notion.so/Alfred-Driver-Fixes-3cf6e0f18c8349fcbc3c880670d6b9a2?pvs=4

**Comment by praveenVnktsh** [comment]
Final FVD Code added, bunch of changes. To be squashed and merged.

**Pull request merged**